Remove environment from transport client

rjernst opened the pull request:
Transport clients run embedded within external applications, so
elasticsearch should not be doing anything with the filesystem, as there
is not elasticsearch home.

This change makes a number of cleanups to the internal API for loading
settings and creating an environment. The loadFromConfig option was
removed, since it was always true except for tests. We now always
attempt to load settings from config a file when an environment is
created. The prepare methods were also simplified so there is now
prepareSettingsAndEnvironment which nodes use, and prepareSettings which
the transport client uses. I also attempted to improve the tests, but
there is a still a lot of follow up work to do there.

closes #13155

NOTE: I left a handful of TODOs for things that looked fishy in prepareSettingsAndEnvironment. In general this should really be cleaned up to be much simpler. There are far to many ways to configure the settings for a developer to understand the code, so I don't know how a user would understand whether to put a setting in properties, env vars, or one of many possible files, or forcing a property, or setting a default, etc.

s1monw commented on core/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java:
```diff
@@ -197,7 +197,7 @@ private static void setupLogging(Settings settings, Environment environment) {
     private static Tuple<Settings, Environment> initialSettings(boolean foreground) {
         Terminal terminal = foreground ? Terminal.DEFAULT : null;
-        return InternalSettingsPreparer.prepareSettings(EMPTY_SETTINGS, true, terminal);
+        return InternalSettingsPreparer.prepareSettingsAndEnvironment(EMPTY_SETTINGS, terminal);
```
it might be simpler to just name this `InternalSettingsPreparer.prepareEnvironment` and return only `Environment` since it has the settings already as a member then we can spare the tuple and things are less funky?

rjernst replied:
I think the only issue with that right now is the last line of the method, which modifies the log file setting to be absolute, after creating the environment.

s1monw replied:
yeah I wonder if we should do this first?

rjernst replied:
well it is using the resolved config dir from the environment...

rjernst replied:
but i agree, this is already breaking. i will move it and simplify as you suggested

s1monw commented on core/src/main/java/org/elasticsearch/node/internal/InternalSettingsPreparer.java:
```diff
@@ -52,22 +53,22 @@
 public class InternalSettingsPreparer {
 
-    static final List<String> ALLOWED_SUFFIXES = Arrays.asList(".yml", ".yaml", ".json", ".properties");
+    static final String[] ALLOWED_SUFFIXES = {".yml", ".yaml", ".json", ".properties"};
```
should these be private? not sure who is using it...

s1monw commented on core/src/main/java/org/elasticsearch/node/internal/InternalSettingsPreparer.java:
```diff
@@ -75,126 +76,143 @@
-                    loadFromEnv = false;
-                    settingsBuilder.loadFromPath(environment.configFile().resolve(System.getProperty("elasticsearch.config")));
-                }
+        // TODO: can we simplify all of this and have a single filename, which is looked up in the config dir?
```
huge +1

s1monw commented on core/src/main/java/org/elasticsearch/node/internal/InternalSettingsPreparer.java:
```diff
@@ -220,37 +238,40 @@ static String randomNodeName(Environment environment) {
+            if (Strings.hasLength(textValue)) {
+                settings.put(setting, textValue);
+            } else {
+                // TODO: why do we remove settings if prompt returns empty??
```
this deserves a sep issue I guess but good catch

s1monw commented on core/src/main/java/org/elasticsearch/plugins/PluginsService.java:
```diff
@@ -87,10 +87,10 @@
      * @param classpathPlugins Plugins that exist in the classpath which should be loaded
      */
-    public PluginsService(Settings settings, Environment environment, Collection<Class<? extends Plugin>> classpathPlugins) {
+    public PluginsService(Settings settings, Path pluginsDirectory, Collection<Class<? extends Plugin>> classpathPlugins) {
```
can we now remove the import `import org.elasticsearch.env.Environment;` here?

s1monw commented:
left some minors LGTM otherwise

clintongormley commented:
@rjernst could you also add a note to the breaking changes docs please: https://www.elastic.co/guide/en/elasticsearch/reference/2.0/_java_api_changes.html

s1monw commented:
@clintongormley what is the breaking change here?

clintongormley commented:
@s1monw the fact that the node client now requires a `path.home` setting, which it didn't before.

rjernst commented:
It's also breaking because the `loadConfigSettings` argument is gone.

@clintongormley I added a note to the migration guide.

rjernst commented:
@s1monw I updated to remove the extra path.logs setter at the end of prepare. It was only there for logging, so I moved it directly to the logging configurator.

s1monw commented:
LGTM

rjernst commented:
2.x commit: 2d698d6
2.0 commit: 5cd65d0

mtraynham commented:
Just ran into this with an upgrade from 1.7 to 2.2, since we were using `elasticsearch.yml` to config the node name for our TransportClient.

I want to point out that the documentation ([here](https://www.elastic.co/guide/en/elasticsearch/client/java-api/current/transport-client.html)) is wrong to suggest you can still do this.

> Or using elasticsearch.yml file as shown in Node Client

clintongormley commented:
thanks @mtraynham - fixed in https://github.com/elastic/elasticsearch/commit/1769e9b78a5f8c17aa5ca5d8f00916a6ac96cfb8
